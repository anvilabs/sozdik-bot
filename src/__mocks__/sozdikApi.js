/* @flow */

import _ from 'lodash/fp';

import type {LanguageType} from '../sozdikApi';

const sampleTranslation = {
  query: 'query',
  text: 'text',
  title: 'title',
  url: 'url',
};

const translations = {
  // prettier-ignore
  'машина': [
    {
      ...sampleTranslation,
      query: 'машина',
      fromLang: 'kk',
      toLang: 'ru',
    },
    {
      ...sampleTranslation,
      query: 'машина',
      fromLang: 'ru',
      toLang: 'kk',
    },
  ],
  // prettier-ignore
  'словарь': [
    {
      ...sampleTranslation,
      query: 'словарь',
      fromLang: 'ru',
      toLang: 'kk',
    },
  ],
  // prettier-ignore
  'лұғат': [
    {
      ...sampleTranslation,
      query: 'лұғат',
      fromLang: 'kk',
      toLang: 'ru',
    },
  ],
};

const sozdikApi = () => ({
  getTranslation: jest.fn(
    (query: string, fromLang: LanguageType, toLang: LanguageType) =>
      _.find({toLang, fromLang}, translations[query]),
  ),
  getTranslationsForQuery: jest.fn(
    (query: string) => translations[query] || [],
  ),
});

export default sozdikApi;
