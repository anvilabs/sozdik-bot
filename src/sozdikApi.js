/* @flow */

import crypto from 'crypto';

import _ from 'lodash/fp';
import rp from 'request-promise';
import toMarkdown from 'to-markdown';

import createLogger from 'src/createLogger';
import config from 'src/config';

export type LanguageType = 'ru' | 'kk';
export type TranslationType = {
  toLang: LanguageType,
  fromLang: LanguageType,
  text: string,
  url: string,
  title: string,
};
export type GetTranslationFnType = (
  query: string,
  fromLang: LanguageType,
  toLang: LanguageType,
) => Promise<?TranslationType>;
export type GetTranslationForQueryFnType = (
  query: string,
) => Promise<Array<TranslationType>>;

const logger = createLogger('sozdik-api');

const formattedLanguageMappings = {
  ru: 'по-русски',
  kk: 'по-казахски',
};
const request = rp.defaults({
  headers: {'User-Agent': 'sozdik-bot'},
  gzip: true,
  json: true,
});

const sozdikApi = (client: 'telegram' | 'facebook') => {
  const apiKey = config.sozdikApiKey[client];

  const getTranslation = async (
    query: string,
    fromLang: LanguageType,
    toLang: LanguageType,
  ) => {
    const hash = crypto
      .createHash('md5')
      .update(`${client}${apiKey}${fromLang}${toLang}${query}`, 'utf8')
      .digest('hex');

    const json = await request({
      url: 'http://api.sozdik.kz/translate',
      qs: {
        client,
        lang_from: fromLang,
        lang_to: toLang,
        phrase: query,
        hash,
        strict: 0,
        output_format: 'json',
        output_samples: 1,
        api_version: '1.0',
        client_version: '0.1.0',
      },
    });

    if (json.message !== 'Found' || !json.translation) return null;

    const converters = [
      {
        filter: 'abbr',
        replacement: (content: string) => `_${content}_`,
      },
      {
        filter: (node: any) =>
          node.nodeName === 'A' &&
          node.getAttribute('href') &&
          !/^https?:\/\//.test(node.getAttribute('href')),
        replacement: (content: string, node: any) =>
          `[${content}](https://sozdik.kz/ru${node.getAttribute('href')}` +
          `${node.title ? ` "${node.title}"` : ''})`,
      },
    ];

    const text = toMarkdown(json.translation, {converters})
      .replace(/(\d+)\\./g, '$1.')
      .replace(/<\/?span>/g, '');

    if (!text) return null;

    const translation: TranslationType = {
      query,
      text,
      fromLang,
      toLang,
      url: json.url_short,
      title: `*"${query}" ${formattedLanguageMappings[toLang]}*`,
    };

    return translation;
  };

  const getTranslationsForQuery = async (query: string) => {
    const translations: Array<TranslationType> = _.compact(
      await Promise.all([
        getTranslation(query, 'kk', 'ru'),
        !/[әіңғүұқөһ]/i.test(query) && getTranslation(query, 'ru', 'kk'),
      ]),
    );

    if (translations.length > 0) {
      _.forEach((translation: TranslationType) => {
        logger.debug(
          `Found a ${translation.toLang === 'ru' ? 'Russian' : 'Kazakh'} ` +
            `translation for "${query}"`,
        );
      }, translations);
    } else {
      logger.debug(`No translations found for "${query}"`);
    }

    return translations;
  };

  return {getTranslation, getTranslationsForQuery};
};

export default sozdikApi;
export {logger as __logger};
