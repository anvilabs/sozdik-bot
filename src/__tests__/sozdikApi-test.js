/* @flow */

import sozdikApi from '../sozdikApi';

jest.unmock('request-promise');
jest.unmock('../sozdikApi');

const {
  getTranslation,
  getTranslationsForQuery,
} = sozdikApi('facebook');

describe('sozdikApi', () => {
  it('translates from ru to kk', async () => {
    expect(await getTranslation('словарь', 'ru', 'kk')).toMatchSnapshot();
  });

  it('translates from kk to ru', async () => {
    expect(await getTranslation('лұғат', 'kk', 'ru')).toMatchSnapshot();
  });

  it('fails to translate an unknown word', async () => {
    expect(await getTranslation('блаблабла', 'ru', 'kk')).toBeNull();
  });

  it('returns all translations for an universal word', async () => {
    expect(await getTranslationsForQuery('машина')).toMatchSnapshot();
  });

  it('returns only a kk translation for a kk word', async () => {
    expect(await getTranslationsForQuery('лұғат')).toMatchSnapshot();
  });

  it('returns only a ru translation for a ru word', async () => {
    expect(await getTranslationsForQuery('словарь')).toMatchSnapshot();
  });
});
