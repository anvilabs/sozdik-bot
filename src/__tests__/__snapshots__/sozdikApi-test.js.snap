// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sozdikApi returns all translations for an universal word 1`] = `
Array [
  Object {
    "fromLang": "kk",
    "query": "машина",
    "text": "1. [машина](https://sozdik.kz/ru/dictionary/translate/ru/kk/машина/)  
2. [машинный](https://sozdik.kz/ru/dictionary/translate/ru/kk/машинный/)",
    "title": "*\\"машина\\" по-русски*",
    "toLang": "ru",
    "url": "http://szd.kz/DAOGLy",
  },
  Object {
    "fromLang": "ru",
    "query": "машина",
    "text": "_жен._ _тех._ [машина](https://sozdik.kz/ru/dictionary/translate/kk/ru/машина/) _разг._ _(автомобиль)_ [көлік](https://sozdik.kz/ru/dictionary/translate/kk/ru/көлік/); [машина](https://sozdik.kz/ru/dictionary/translate/kk/ru/машина/) _перен._ [машина](https://sozdik.kz/ru/dictionary/translate/kk/ru/машина/)",
    "title": "*\\"машина\\" по-казахски*",
    "toLang": "kk",
    "url": "http://szd.kz/wbwRZc",
  },
]
`;

exports[`sozdikApi returns only a kk translation for a kk word 1`] = `
Array [
  Object {
    "fromLang": "kk",
    "query": "лұғат",
    "text": "_ар._ _уст._ [язык](https://sozdik.kz/ru/dictionary/translate/ru/kk/язык/) _книжн._ [словарь](https://sozdik.kz/ru/dictionary/translate/ru/kk/словарь/)  
_син._ [сөздік](https://sozdik.kz/ru/dictionary/translate/kk/ru/сөздік/), [сөзтізбе](https://sozdik.kz/ru/dictionary/translate/kk/ru/сөзтізбе/)",
    "title": "*\\"лұғат\\" по-русски*",
    "toLang": "ru",
    "url": "http://szd.kz/*PdASd",
  },
]
`;

exports[`sozdikApi returns only a ru translation for a ru word 1`] = `
Array [
  Object {
    "fromLang": "ru",
    "query": "словарь",
    "text": "_муж._ [сөздік](https://sozdik.kz/ru/dictionary/translate/kk/ru/сөздік/); [лұғат](https://sozdik.kz/ru/dictionary/translate/kk/ru/лұғат/) _только ед._ _(лексика)_ сөз байлығы; сөздік құрамы _(бір тілдің не жазушының)_",
    "title": "*\\"словарь\\" по-казахски*",
    "toLang": "kk",
    "url": "http://szd.kz/Misd56",
  },
]
`;

exports[`sozdikApi translates from kk to ru 1`] = `
Object {
  "fromLang": "kk",
  "query": "лұғат",
  "text": "_ар._ _уст._ [язык](https://sozdik.kz/ru/dictionary/translate/ru/kk/язык/) _книжн._ [словарь](https://sozdik.kz/ru/dictionary/translate/ru/kk/словарь/)  
_син._ [сөздік](https://sozdik.kz/ru/dictionary/translate/kk/ru/сөздік/), [сөзтізбе](https://sozdik.kz/ru/dictionary/translate/kk/ru/сөзтізбе/)",
  "title": "*\\"лұғат\\" по-русски*",
  "toLang": "ru",
  "url": "http://szd.kz/*PdASd",
}
`;

exports[`sozdikApi translates from ru to kk 1`] = `
Object {
  "fromLang": "ru",
  "query": "словарь",
  "text": "_муж._ [сөздік](https://sozdik.kz/ru/dictionary/translate/kk/ru/сөздік/); [лұғат](https://sozdik.kz/ru/dictionary/translate/kk/ru/лұғат/) _только ед._ _(лексика)_ сөз байлығы; сөздік құрамы _(бір тілдің не жазушының)_",
  "title": "*\\"словарь\\" по-казахски*",
  "toLang": "kk",
  "url": "http://szd.kz/Misd56",
}
`;
